qiskit/pulse/builder.py

```diff
@@ -773,39 +790,47 @@ def call_subroutine(
             self._compile_lazy_circuit()
             subroutine = self._compile_circuit(subroutine)
 
-        empty_subroutine = True
-        if isinstance(subroutine, Schedule):
-            if len(subroutine.instructions) > 0:
-                empty_subroutine = False
-        elif isinstance(subroutine, ScheduleBlock):
-            if len(subroutine.blocks) > 0:
-                empty_subroutine = False
-        else:
+        if not isinstance(subroutine, (Schedule, ScheduleBlock)):
             raise exceptions.PulseError(
                 f"Subroutine type {subroutine.__class__.__name__} is "
                 "not valid data format. Call QuantumCircuit, "
                 "Schedule, or ScheduleBlock."
             )
 
-        if not empty_subroutine:
-            param_value_map = {}
-            for param_name, assigned_value in kw_params.items():
-                param_objs = subroutine.get_parameters(param_name)
-                if len(param_objs) > 0:
-                    for param_obj in param_objs:
-                        param_value_map[param_obj] = assigned_value
-                else:
-                    raise exceptions.PulseError(
-                        f"Parameter {param_name} is not defined in the target subroutine. "
-                        f'{", ".join(map(str, subroutine.parameters))} can be specified.'
-                    )
-
-            if value_dict:
-                param_value_map.update(value_dict)
+        if len(subroutine) == 0:
+            return
 
-            call_def = instructions.Call(subroutine, param_value_map, name)
-
-            self.append_instruction(call_def)
+        # Create local parameter assignment
+        local_assignment = dict()
+        for param_name, value in kw_params.items():
+            params = subroutine.get_parameters(param_name)
+            if not params:
+                raise exceptions.PulseError(
+                    f"Parameter {param_name} is not defined in the target subroutine. "
+                    f'{", ".join(map(str, subroutine.parameters))} can be specified.'
+                )
+            for param in params:
+                local_assignment[param] = value
+        if value_dict:
+            local_assignment.update(value_dict)
+
+        sub_name = name or subroutine.name
+
+        if isinstance(subroutine, ScheduleBlock):
+            # If subroutine is schedule block, use reference mechanism.
```

Comment:
I'm thinking to remove the use case of calling block as a `Call` instruction. I don't think there are many users directly building a pulse program without builder syntax, so this will be kind of internal mechanism update. 

The reason `Call` is introduced is to improve memory efficiency at program construction and serialization. A `Call` instance separately stores parameter-unassigned schedule and parameter table, and it is intended to reuse unassigned schedule among multiple instances. However, it is difficult to perform deduplication of schedules stored in separate instruction instances, on the other hand, reference mechanism easily realizes this because subroutine is now stored in the schedule block. 
Reference mechanism doesn't separately keep parameter-unassigned schedule, but I don't have any use case in mind that requires assignment of different parameters to the same parameterized sequence within the same schedule (i.e. Rabi experiment scans pulse amp, but each scan is different schedule instance). Indeed, current `Call` just degrades memory efficiency by having two schedule instances (one unassigned and other assigned one in cache) within is instance.

If we keep calling the schedule block, then called schedule is not recognized as a reference. If a program has mixed representation of call and reference, it seems like we can easily run into some edge case.